frontend/src/metabase/common/components/QueryColumnPicker/QueryColumnPicker.tsx

```diff
@@ -48,26 +59,42 @@ export function QueryColumnPicker({
   color = "brand",
   checkIsColumnSelected,
   onSelect,
+  onSelectSegment,
   onClose,
 }: QueryColumnPickerProps) {
-  const sections: Sections[] = useMemo(
-    () =>
-      columnGroups.map(group => {
+  const sections: Sections[] = useMemo(() => {
+    const columnGroupSections: Sections[] = columnGroups.map(
+      (group, groupIndex) => {
         const groupInfo = Lib.displayInfo(query, stageIndex, group);
 
-        const items = Lib.getColumnsFromColumnGroup(group).map(column => ({
-          ...Lib.displayInfo(query, stageIndex, column),
-          column,
-        }));
+        const columnItems = Lib.getColumnsFromColumnGroup(group).map(
+          column => ({
+            ...Lib.displayInfo(query, stageIndex, column),
+            column,
+          }),
+        );
+
+        const includeSegments = !!onSelectSegment && !groupInfo.isFromJoin;
+
+        const segmentItems = includeSegments
+          ? Lib.availableSegments(query).map(
+              (segment: Lib.SegmentMetadata) => ({
+                ...Lib.displayInfo(query, stageIndex, segment),
+                segment,
+              }),
+            )
+          : [];
```

Comment:
I don't *love* overloading the columnPicker to be a column-or-segment picker, but it seemed insane to create a separate picker that duplicated all of the column picker logic. 

I wonder if it's worth reworking the columnPicker to be more generic? I also fear that might be grossly over-engineering it.